source/extensions/common/proxy_protocol/proxy_protocol_header.cc

```diff
@@ -111,12 +111,36 @@ void generateV2Header(const Network::Address::Ip& source_address,
 }
 
 bool generateV2Header(const Network::ProxyProtocolData& proxy_proto_data, Buffer::Instance& out,
-                      bool pass_all_tlvs, const absl::flat_hash_set<uint8_t>& pass_through_tlvs) {
-  uint64_t extension_length = 0;
-  for (auto&& tlv : proxy_proto_data.tlv_vector_) {
+                      bool pass_all_tlvs, const absl::flat_hash_set<uint8_t>& pass_through_tlvs,
+                      const std::vector<Envoy::Network::ProxyProtocolTLV>& custom_tlvs) {
+  std::vector<Envoy::Network::ProxyProtocolTLV> combined_tlv_vector;
+  combined_tlv_vector.reserve(custom_tlvs.size() + proxy_proto_data.tlv_vector_.size());
+
+  absl::flat_hash_set<uint8_t> seen_types;
+  for (const auto& tlv : custom_tlvs) {
+    if (seen_types.contains(tlv.type)) {
+      ENVOY_LOG_MISC(warn, "Ignoring duplicate custom TLV type {}", tlv.type);
```

Comment:
The check for duplicates should be done while loading the config (and generate a configuration error for that case). Then it can be an ASSERT here.

Reply:
Good call. I have the ASSERT implementation staged locally but running into some issues with the corresponding test suite:

```cpp
  absl::flat_hash_set<uint8_t> seen_types;
  for (const auto& tlv : custom_tlvs) {
    ASSERT(!seen_types.contains(tlv.type));
    combined_tlv_vector.emplace_back(tlv);
    seen_types.insert(tlv.type);
  }
```

I assume I need to use ASSERT_FALSE when the generateV2Header function is called in the common header package unless I'm missing something obvious? Right now that suite is complaining about segfault, but it wasn't immediately clear to me whether that's the expected behavior or whether the refactored code I have staged locally requires further investigation. I'll assume it's the latter and step through tomorrow.

Reply:
Done in https://github.com/envoyproxy/envoy/pull/37591/commits/c5a5619c693e09c52760d7695f907b00ea109ecb.